File: hackerrank/medium/0008-bigger-is-greater/python/solution.py
## solution.py

from typing import List


def biggerIsGreater(w: str) -> str:
    chars: List[str] = list(w)
    n = len(chars)
    if n < 2:
        return "no answer"

    # Step 1: find pivot i
    i = n - 2
    while i >= 0 and chars[i] >= chars[i + 1]:
        i -= 1

    if i < 0:
        return "no answer"

    # Step 2: find rightmost j > i with chars[j] > chars[i]
    j = n - 1
    while chars[j] <= chars[i]:
        j -= 1

    # Step 3: swap chars[i], chars[j]
    chars[i], chars[j] = chars[j], chars[i]

    # Step 4: reverse suffix from i+1 to end
    chars[i + 1 :] = reversed(chars[i + 1 :])

    return "".join(chars)


# --- demo run ---
if __name__ == "__main__":
    tests = [
        "ab",
        "bb",
        "hefg",
        "dhck",
        "dkhc",
        "lmno",
        "dcba",
        "dcbb",
        "abdc",
        "abcd",
        "fedcbabcd",
    ]

    for t in tests:
        print(f"{t} -> {biggerIsGreater(t)}")

File: hackerrank/medium/0008-bigger-is-greater/js/solution.mjs
// solution.mjs

export function biggerIsGreater(w) {
  const chars = w.split("");
  const n = chars.length;

  if (n < 2) return "no answer";

  // Step 1: find pivot i
  let i = n - 2;
  while (i >= 0 && chars[i] >= chars[i + 1]) {
    i--;
  }
  if (i < 0) {
    return "no answer";
  }

  // Step 2: find rightmost j > i with chars[j] > chars[i]
  let j = n - 1;
  while (chars[j] <= chars[i]) {
    j--;
  }

  // Step 3: swap chars[i], chars[j]
  [chars[i], chars[j]] = [chars[j], chars[i]];

  // Step 4: reverse suffix from i+1 to end
  let left = i + 1;
  let right = n - 1;
  while (left < right) {
    [chars[left], chars[right]] = [chars[right], chars[left]];
    left++;
    right--;
  }

  return chars.join("");
}

// --- demo run ---
if (import.meta.url === `file://${process.argv[1]}`) {
  const tests = [
    "ab",
    "bb",
    "hefg",
    "dhck",
    "dkhc",
    "lmno",
    "dcba",
    "dcbb",
    "abdc",
    "abcd",
    "fedcbabcd",
  ];

  for (const t of tests) {
    console.log(`${t} -> ${biggerIsGreater(t)}`);
  }
}
